libfwupdplugin/fu-archive-firmware.c

```diff
@@ -46,14 +51,45 @@ fu_archive_firmware_parse(FuFirmware *firmware,
 	g_autoptr(FuArchive) archive = NULL;
 
 	/* load archive */
-	archive = fu_archive_new(fw, FU_ARCHIVE_FLAG_IGNORE_PATH, error);
+	archive = fu_archive_new_from_bytes(fw, FU_ARCHIVE_FLAG_IGNORE_PATH, error);
 	if (archive == NULL)
 		return FALSE;
 
 	/* decompress each image in the archive */
 	return fu_archive_iterate(archive, fu_archive_firmware_parse_cb, firmware, error);
 }
 
+/* TODO: Create an FuAchiveEntry to replace the filename by an entry structure? */
+static GBytes *
+fu_archive_firmware_write_cb(FuArchive *self,
+			     const gchar **filename,
+			     gpointer user_data,
+			     GError **error)
+{
+	FuArchiveFirmwareIter *iter = (FuArchiveFirmwareIter *)user_data;
+	if (iter->index < iter->array->len) {
+		FuFirmware *img = g_ptr_array_index(iter->array, iter->index++);
+		*filename =  fu_firmware_get_id(img);
+		return fu_firmware_get_bytes(img, error);
+	}
+	*filename = NULL;
+	return g_bytes_new(NULL, 0);
```

Comment:
I do no really like returning an empty `GBytes` to stop.

Any better idea?

Reply:
Maybe a gboolean in the `FuArchiveFirmwareIter` of `done` or something?